### ISSUE TITLE
TypeError: htmlspecialchars(): Argument #1 ($string) must be of type string, array given in \public_html\vendor\october\rain\src\Support\helpers.php:261

### ISSUE THREAD

**Issue by dathwa** [rainlab/user-plugin#549]
Hello.
OCMS 3.6.24
Just did the big Users update from 2.0 to 3.0.7
Navigating to `/backend/system/settings/update/rainlab/user/settings` and i get this: `TypeError: htmlspecialchars(): Argument #1 ($string) must be of type string, array given in \public_html\vendor\october\rain\src\Support\helpers.php:261`

Can you help? Thanks

```
[2024-06-20 11:15:40] development.ERROR: TypeError: htmlspecialchars(): Argument #1 ($string) must be of type string, array given in \public_html\vendor\october\rain\src\Support\helpers.php:261
Stack trace:
#0 \public_html\vendor\october\rain\src\Support\helpers.php(261): htmlspecialchars(Array, 3, 'UTF-8', false)
#1 \public_html\modules\backend\classes\FormField.php(315): e(Array)
#2 \public_html\modules\backend\widgets\form\partials\_field_dropdown.php(58): Backend\Classes\FormField->getDisplayValue('offline.mall::m...')
#3 \public_html\modules\system\traits\ViewMaker.php(272): include('C:\\htdocs\\aps.o...')
#4 \public_html\modules\system\traits\ViewMaker.php(103): Backend\Classes\WidgetBase->makeFileContents('C:\\htdocs\\aps.o...', Array)
#5 \public_html\modules\backend\widgets\Form.php(343): Backend\Classes\WidgetBase->makePartial('_field_dropdown', Array)
#6 \public_html\modules\backend\widgets\form\partials\_field.php(25): Backend\Widgets\Form->renderFieldElement(Object(Backend\Classes\FormField))
#7 \public_html\modules\system\traits\ViewMaker.php(272): include('C:\\htdocs\\aps.o...')
#8 \public_html\modules\system\traits\ViewMaker.php(103): Backend\Classes\WidgetBase->makeFileContents('C:\\htdocs\\aps.o...', Array)
#9 \public_html\modules\backend\widgets\form\partials\_field-container.php(10): Backend\Classes\WidgetBase->makePartial('_field', Array)
#10 \public_html\modules\system\traits\ViewMaker.php(272): include('C:\\htdocs\\aps.o...')
#11 \public_html\modules\system\traits\ViewMaker.php(103): Backend\Classes\WidgetBase->makeFileContents('C:\\htdocs\\aps.o...', Array)
#12 \public_html\modules\backend\widgets\form\partials\_form_fields.php(11): Backend\Classes\WidgetBase->makePartial('_field-containe...', Array)
#13 \public_html\modules\system\traits\ViewMaker.php(272): include('C:\\htdocs\\aps.o...')
#14 \public_html\modules\system\traits\ViewMaker.php(103): Backend\Classes\WidgetBase->makeFileContents('C:\\htdocs\\aps.o...', Array)
#15 \public_html\modules\backend\widgets\form\partials\_form_tabs.php(63): Backend\Classes\WidgetBase->makePartial('_form_fields', Array)
#16 \public_html\modules\system\traits\ViewMaker.php(272): include('C:\\htdocs\\aps.o...')
#17 \public_html\modules\system\traits\ViewMaker.php(103): Backend\Classes\WidgetBase->makeFileContents('C:\\htdocs\\aps.o...', Array)
#18 \public_html\modules\backend\widgets\form\partials\_section.php(28): Backend\Classes\WidgetBase->makePartial('_form_tabs', Array)
#19 \public_html\modules\system\traits\ViewMaker.php(272): include('C:\\htdocs\\aps.o...')
#20 \public_html\modules\system\traits\ViewMaker.php(103): Backend\Classes\WidgetBase->makeFileContents('C:\\htdocs\\aps.o...', Array)
#21 \public_html\modules\backend\widgets\form\partials\_form.php(6): Backend\Classes\WidgetBase->makePartial('_section', Array)
#22 \public_html\modules\system\traits\ViewMaker.php(272): include('C:\\htdocs\\aps.o...')
#23 \public_html\modules\system\traits\ViewMaker.php(103): Backend\Classes\WidgetBase->makeFileContents('C:\\htdocs\\aps.o...', Array)
#24 \public_html\modules\backend\widgets\form\partials\_form-container.php(7): Backend\Classes\WidgetBase->makePartial('_form')
#25 \public_html\modules\system\traits\ViewMaker.php(272): include('C:\\htdocs\\aps.o...')
#26 \public_html\modules\system\traits\ViewMaker.php(103): Backend\Classes\WidgetBase->makeFileContents('C:\\htdocs\\aps.o...', Array)
#27 \public_html\modules\backend\widgets\Form.php(246): Backend\Classes\WidgetBase->makePartial('_form-container', Array)
#28 \public_html\modules\system\controllers\Settings.php(160): Backend\Widgets\Form->render(Array)
#29 \public_html\modules\system\controllers\settings\update.php(4): System\Controllers\Settings->formRender()
#30 \public_html\modules\system\traits\ViewMaker.php(272): include('C:\\htdocs\\aps.o...')
#31 \public_html\modules\system\traits\ViewMaker.php(117): Backend\Classes\Controller->makeFileContents('C:\\htdocs\\aps.o...')
#32 \public_html\modules\backend\classes\Controller.php(406): Backend\Classes\Controller->makeView('update')
#33 \public_html\modules\backend\classes\Controller.php(274): Backend\Classes\Controller->execPageAction('update', Array)
#34 \public_html\modules\backend\classes\BackendController.php(98): Backend\Classes\Controller->run('update', Array)
#35 \public_html\vendor\laravel\framework\src\Illuminate\Routing\Controller.php(54): Backend\Classes\BackendController->run('system/settings...')
#36 \public_html\vendor\laravel\framework\src\Illuminate\Routing\ControllerDispatcher.php(43): Illuminate\Routing\Controller->callAction('run', Array)
#37 \public_html\vendor\laravel\framework\src\Illuminate\Routing\Route.php(260): Illuminate\Routing\ControllerDispatcher->dispatch(Object(Illuminate\Routing\Route), Object(Backend\Classes\BackendController), 'run')
#38 \public_html\vendor\laravel\framework\src\Illuminate\Routing\Route.php(205): Illuminate\Routing\Route->runController()
#39 \public_html\vendor\laravel\framework\src\Illuminate\Routing\Router.php(798): Illuminate\Routing\Route->run()
#40 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(141): Illuminate\Routing\Router->Illuminate\Routing\{closure}(Object(Illuminate\Http\Request))
#41 \public_html\vendor\laravel\framework\src\Illuminate\Routing\Middleware\SubstituteBindings.php(50): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#42 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(180): Illuminate\Routing\Middleware\SubstituteBindings->handle(Object(Illuminate\Http\Request), Object(Closure))
#43 \public_html\vendor\laravel\framework\src\Illuminate\View\Middleware\ShareErrorsFromSession.php(49): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#44 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(180): Illuminate\View\Middleware\ShareErrorsFromSession->handle(Object(Illuminate\Http\Request), Object(Closure))
#45 \public_html\vendor\laravel\framework\src\Illuminate\Session\Middleware\StartSession.php(121): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#46 \public_html\vendor\laravel\framework\src\Illuminate\Session\Middleware\StartSession.php(64): Illuminate\Session\Middleware\StartSession->handleStatefulRequest(Object(Illuminate\Http\Request), Object(Illuminate\Session\Store), Object(Closure))
#47 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(180): Illuminate\Session\Middleware\StartSession->handle(Object(Illuminate\Http\Request), Object(Closure))
#48 \public_html\vendor\laravel\framework\src\Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse.php(37): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#49 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(180): Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse->handle(Object(Illuminate\Http\Request), Object(Closure))
#50 \public_html\vendor\laravel\framework\src\Illuminate\Cookie\Middleware\EncryptCookies.php(67): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#51 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(180): Illuminate\Cookie\Middleware\EncryptCookies->handle(Object(Illuminate\Http\Request), Object(Closure))
#52 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(116): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#53 \public_html\vendor\laravel\framework\src\Illuminate\Routing\Router.php(799): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#54 \public_html\vendor\laravel\framework\src\Illuminate\Routing\Router.php(776): Illuminate\Routing\Router->runRouteWithinStack(Object(Illuminate\Routing\Route), Object(Illuminate\Http\Request))
#55 \public_html\vendor\laravel\framework\src\Illuminate\Routing\Router.php(740): Illuminate\Routing\Router->runRoute(Object(Illuminate\Http\Request), Object(Illuminate\Routing\Route))
#56 \public_html\vendor\october\rain\src\Router\CoreRouter.php(32): Illuminate\Routing\Router->dispatchToRoute(Object(Illuminate\Http\Request))
#57 \public_html\vendor\laravel\framework\src\Illuminate\Foundation\Http\Kernel.php(190): October\Rain\Router\CoreRouter->dispatch(Object(Illuminate\Http\Request))
#58 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(141): Illuminate\Foundation\Http\Kernel->Illuminate\Foundation\Http\{closure}(Object(Illuminate\Http\Request))
#59 \public_html\plugins\offline\speedy\classes\middleware\CDNMiddleware.php(23): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#60 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(180): OFFLINE\Speedy\Classes\Middleware\CDNMiddleware->handle(Object(Illuminate\Http\Request), Object(Closure))
#61 \public_html\plugins\offline\speedy\classes\middleware\Http2Middleware.php(38): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#62 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(180): OFFLINE\Speedy\Classes\Middleware\Http2Middleware->handle(Object(Illuminate\Http\Request), Object(Closure))
#63 \public_html\plugins\offline\responsiveimages\classes\ResponsiveImagesMiddleware.php(30): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#64 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(180): OFFLINE\ResponsiveImages\Classes\ResponsiveImagesMiddleware->handle(Object(Illuminate\Http\Request), Object(Closure))
#65 \public_html\vendor\laravel\framework\src\Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance.php(86): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#66 \public_html\vendor\october\rain\src\Foundation\Http\Middleware\CheckForMaintenanceMode.php(23): Illuminate\Foundation\Http\Middleware\PreventRequestsDuringMaintenance->handle(Object(Illuminate\Http\Request), Object(Closure))
#67 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(180): October\Rain\Foundation\Http\Middleware\CheckForMaintenanceMode->handle(Object(Illuminate\Http\Request), Object(Closure))
#68 \public_html\vendor\laravel\framework\src\Illuminate\Pipeline\Pipeline.php(116): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline\{closure}(Object(Illuminate\Http\Request))
#69 \public_html\vendor\laravel\framework\src\Illuminate\Foundation\Http\Kernel.php(165): Illuminate\Pipeline\Pipeline->then(Object(Closure))
#70 \public_html\vendor\laravel\framework\src\Illuminate\Foundation\Http\Kernel.php(134): Illuminate\Foundation\Http\Kernel->sendRequestThroughRouter(Object(Illuminate\Http\Request))
#71 \public_html\index.php(43): Illuminate\Foundation\Http\Kernel->handle(Object(Illuminate\Http\Request))
#72 {main}  

```

**Comment by dathwa** [comment]
Did a trace_log on line 261 for $value and got below.
Strange. this is from the Mall plugin. 

```
[2024-06-20 13:05:18] development.INFO: offline.mall::mail.admin.checkout_failed  
[2024-06-20 13:05:18] development.INFO: Array
(
    [subject] => Checkout failed #:number
    [not_processed] => The following order could not be processed correctly. It is possible that you have to contact the customer.
    [error_details] => Error message
    [payment_id] => Payment ID
    [error] => Error details
)
```

**Comment by dathwa** [comment]
Seems to be these 2 field settings `plugins/rainlab/user/models/setting/fields.yaml` that are breaking it.

 ```
 user_message_template:
            label: User Message Template
            commentAbove: Select the mail template to send to the user.
            type: dropdown
            optionsMethod: System\Models\MailTemplate::listAllTemplates
            tab: Notifications
            cssClass: field-indent
            trigger:
                action: enable
                field: notify_user
                condition: checked

        system_message_template:
            label: System Message Template
            commentAbove: Select the mail template to send to the admins.
            type: dropdown
            optionsMethod: System\Models\MailTemplate::listAllTemplates
            tab: Notifications
            cssClass: field-indent
            trigger:
                action: enable
                field: notify_system
                condition: checked
```

**Comment by daftspunk** [comment]
I'm not sure why this would happen unless the Mall plugin depends on the settings YAML? Some options are to stay on v2.0 or let them know about the incompatibility.

**Comment by tobias-kuendig** [comment]
`offline.mall::mail.admin.checkout_failed` is the name of one of our mail templates. We register it like so:

```php
    public function registerMailTemplates()
    {
        return [
            'offline.mall::mail.admin.checkout_failed',
        ];
    }
```

But `offline.mall::mail.admin.checkout_failed` is a message translation key as well, it returns an array:

```php
<?php
return [
    'admin' => [
        'checkout_failed' => [
            'subject' => 'Checkout failed #:number',
            'not_processed' => 'The following order could not be processed correctly. It is possible that you have to contact the customer.',
            'error_details' => 'Error message',
            'payment_id' => 'Payment ID',
            'error' => 'Error details',
        ],
   ];
];
```

It looks like this array value is now passed to the translation service, which results in @dathwa's error.

**Comment by dathwa** [comment]
Hello @tobias-kuendig
It seems we have reached a bit of an impasse.
Would you say that `rainlab/user v3` is incompatible with Mall? 
It's all these lang arrays below that break it...

```
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Checkout failed #:number
    [not_processed] => The following order could not be processed correctly. It is possible that you have to contact the customer.
    [error_details] => Error message
    [payment_id] => Payment ID
    [error] => Error details
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => New order #:number
    [order_placed] => The following order was placed in your shop:
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Payment for order #:number successful
    [message] => The previously failed payment for this order succeeded on a subsequent attempt.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Checkout error for order #:number
    [problem_message] => We are very sorry that there was a problem during your checkout process. We will look into the problem and contact you with further information.
    [check_order_status] => To check the status of your order you may log in to our shop at any time.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Confirmation for order #:number
    [thanks_for_order] => Thank you for your order. We have received your payment and your order is now being processed. Please review your order details below. 
    [check_order_status] => You can check the status of your order by visiting the account section of our shop.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Welcome to the APS shop, :firstname
    [confirm_mail] => Welcome to the APS shop! Please click on the button below to confirm your e-mail address.
    [message] => Welcome to the APS shop! You can log in using your e-mail address **:email** and start shopping immediately.
    [possibilities] => Your user account enables you to keep track of open and past orders.
    [button] => Array
        (
            [confirm] => Confirm your e-mail address
            [visit_store] => Visit our shop
        )

)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Your order has been dispatched
    [message] => Your order **#:number** has now been dispatched.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => The status of your order changed
    [message] => We just wanted to let you know that your order **#:number** was updated to the new status: **:state**
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => The payment for your order has failed
    [message] => We just wanted to let you know that the payment for order **#:number** has failed
    [payment_advice] => Please login to your account and try again to pay the order.
    [support] => If you continue to experience problems with payments please contact us.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Thank you for your payment
    [message] => Thank you for your order **#:number**.
    [process_order] => We have successfully received your order and payment which is now being processed. We will be in touch once your order has been dispatched. Thank you for shopping with APS.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Your payment was refunded
    [message] => As agreed we have made an adjustment to your payment on order **#:number**.
    [duration] => Please be aware that it may take up to 5 working days for the refund to be processed.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Checkout failed #:number
    [not_processed] => The following order could not be processed correctly. It is possible that you have to contact the customer.
    [error_details] => Error message
    [payment_id] => Payment ID
    [error] => Error details
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => New order #:number
    [order_placed] => The following order was placed in your shop:
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Payment for order #:number successful
    [message] => The previously failed payment for this order succeeded on a subsequent attempt.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Checkout error for order #:number
    [problem_message] => We are very sorry that there was a problem during your checkout process. We will look into the problem and contact you with further information.
    [check_order_status] => To check the status of your order you may log in to our shop at any time.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Confirmation for order #:number
    [thanks_for_order] => Thank you for your order. We have received your payment and your order is now being processed. Please review your order details below. 
    [check_order_status] => You can check the status of your order by visiting the account section of our shop.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Welcome to the APS shop, :firstname
    [confirm_mail] => Welcome to the APS shop! Please click on the button below to confirm your e-mail address.
    [message] => Welcome to the APS shop! You can log in using your e-mail address **:email** and start shopping immediately.
    [possibilities] => Your user account enables you to keep track of open and past orders.
    [button] => Array
        (
            [confirm] => Confirm your e-mail address
            [visit_store] => Visit our shop
        )

)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Your order has been dispatched
    [message] => Your order **#:number** has now been dispatched.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => The status of your order changed
    [message] => We just wanted to let you know that your order **#:number** was updated to the new status: **:state**
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => The payment for your order has failed
    [message] => We just wanted to let you know that the payment for order **#:number** has failed
    [payment_advice] => Please login to your account and try again to pay the order.
    [support] => If you continue to experience problems with payments please contact us.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Thank you for your payment
    [message] => Thank you for your order **#:number**.
    [process_order] => We have successfully received your order and payment which is now being processed. We will be in touch once your order has been dispatched. Thank you for shopping with APS.
)
  
[2024-07-18 11:12:44] development.INFO: Array
(
    [subject] => Your payment was refunded
    [message] => As agreed we have made an adjustment to your payment on order **#:number**.
    [duration] => Please be aware that it may take up to 5 working days for the refund to be processed.
)
```

**Comment by daftspunk** [comment]
Hey @tobias-kuendig 

Thanks for the detailed report. These collisions are not uncommon, so we usually wrap the value in `[xxx]`. However, more recently we introduced the `valueTrans` property that disables translation on the values.

Fixed in e023b929b74b5702f830728257b4426b356e84eb (v3.1.1)
